Add bq_ipv6_bias and bt_ndt_server queries to the bigquery exporter

stephen-soltesz opened the pull request:
This change adds two new standardSQL queries for the bigquery exporter.

 * bq_ipv6_bias -- per-machine, per-address-type, and per-window scale metrics for measuring bias introduced by IPv6 configuration on M-Lab servers. This will allow us to safely monitor new deployments.
 * bq_ndt_server -- per-machine summary metrics for NDT uploads, downloads, and RTT. This will allow us to observe changes in "normal" behavior over time from NDT server or kernel upgrades.

This change also includes two minor changes to the apply-global-prometheus.sh script to report configs in the event of errors, and an update of the prometheus container version from 1.6.2 to 1.8.2. The 1.8 version includes features for more easily migrating to version 2.0, which uses a different backing store format.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/prometheus-support/131)
<!-- Reviewable:end -->

gfr10598 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 5 files at r1.
Review status: 4 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[config/federation/bigquery/bq_ipv6_bias.sql, line 2 at r1](https://reviewable.io:443/reviews/m-lab/prometheus-support/131#-KzVu7qMOOHKSkwjC5FC:-KzVu7qNi-ihKjlddem9:b-2imsq4) ([raw file](https://github.com/m-lab/prometheus-support/blob/db0808c972271d905f392cb67008a04ce3647906/config/federation/bigquery/bq_ipv6_bias.sql#L2)):*
> ```SQLPL
> -- bq_ipv6_bias collects the number of daily tests broken down by site, address
> -- type an window scale. The query reports data from two days ago according to
> ```

typo: type *and* window scale

---

*[config/federation/bigquery/bq_ipv6_bias.sql, line 13 at r1](https://reviewable.io:443/reviews/m-lab/prometheus-support/131#-KzVt_wEWRaf-pBkQ6CB:-KzVtXKbmXe02ggUvtAb:bejhjjs) ([raw file](https://github.com/m-lab/prometheus-support/blob/db0808c972271d905f392cb67008a04ce3647906/config/federation/bigquery/bq_ipv6_bias.sql#L13)):*
> ```SQLPL
>        WHEN 1 <= web100_log_entry.snap.SndWindScale AND web100_log_entry.snap.SndWindScale <= 2 THEN "ws12"
>        WHEN 3 <= web100_log_entry.snap.SndWindScale AND web100_log_entry.snap.SndWindScale <= 5 THEN "ws345"
>        WHEN 6 <= web100_log_entry.snap.SndWindScale AND web100_log_entry.snap.SndWindScale <= 8 THEN "ws678"
> ```

Looks like 678 accounts for most of the counts.  Consider splitting into separate 6,7,8 counters?

---

*[config/federation/bigquery/bq_ipv6_bias.sql, line 25 at r1](https://reviewable.io:443/reviews/m-lab/prometheus-support/131#-KzVuLp8HE6E1kwANOVp:-KzVuLp9VIzG_D1MeXfB:b-r75fam) ([raw file](https://github.com/m-lab/prometheus-support/blob/db0808c972271d905f392cb67008a04ce3647906/config/federation/bigquery/bq_ipv6_bias.sql#L25)):*
> ```SQLPL
>     -- To guarantee the period queried is up to date (all data collected and
>     -- parsed), we should wait 36hours after the start of day. To also use
>     -- _PARTITIONTIME boundaries, we must look 2 days in the past.
> ```

The constraint seems a bit more subtle.  If you do this at 1:00 am, some of the data will be only 25 hours old.  Do you care?  I expect with the incremental scraper, it should be fine to run this at noon for the previous day, i.e., only 12 hour delay, right?

---

*[config/federation/bigquery/bq_ipv6_bias.sql, line 31 at r1](https://reviewable.io:443/reviews/m-lab/prometheus-support/131#-KzVto3mVK05K4_XcfwQ:-KzVto3mVK05K4_XcfwR:b-ciscom) ([raw file](https://github.com/m-lab/prometheus-support/blob/db0808c972271d905f392cb67008a04ce3647906/config/federation/bigquery/bq_ipv6_bias.sql#L31)):*
> ```SQLPL
>    site, address_type, window_scale
> 
> ORDER BY
> ```

FYI - be aware that ORDER BY can cause memory problems with large number of items.  Definitely no problem here, but good to keep in mind.

---

*[config/federation/bigquery/bq_ndt_server.sql, line 8 at r1](https://reviewable.io:443/reviews/m-lab/prometheus-support/131#-KzVuzeBfPPNwfAMzHng:-KzVuzeBfPPNwfAMzHnh:b-8276ig) ([raw file](https://github.com/m-lab/prometheus-support/blob/db0808c972271d905f392cb67008a04ce3647906/config/federation/bigquery/bq_ndt_server.sql#L8)):*
> ```SQLPL
> --   bq_ndt_server_tests
> --   bq_ndt_server_download_avg_rate
> --   bq_ndt_server_download_median_rate
> ```

I think that 99th percentile would probably be very useful as well.  Possibly 99.9th?

BTW, found this, which looks interesting:
https://cloud.google.com/blog/big-data/2016/10/a-google-sre-explores-github-reliability-with-bigquery

---

*[config/federation/bigquery/bq_ndt_server.sql, line 12 at r1](https://reviewable.io:443/reviews/m-lab/prometheus-support/131#-KzVyirqrjI8fGUerTBg:-KzVyirqrjI8fGUerTBh:b1lifhq) ([raw file](https://github.com/m-lab/prometheus-support/blob/db0808c972271d905f392cb67008a04ce3647906/config/federation/bigquery/bq_ndt_server.sql#L12)):*
> ```SQLPL
> --   bq_ndt_server_upload_avg_rate
> --   bq_ndt_server_upload_median_rate
> --   bq_ndt_server_upload_median_rtt
> ```

FYI, the #standardSQL only seems to be effective if it is the first line, so recommend putting it at the top.

---

*[config/federation/bigquery/bq_ndt_server.sql, line 19 at r1](https://reviewable.io:443/reviews/m-lab/prometheus-support/131#-KzW-MW2uN7beHJTj_GC:-KzW-MW2uN7beHJTj_GD:bjs8n6a) ([raw file](https://github.com/m-lab/prometheus-support/blob/db0808c972271d905f392cb67008a04ce3647906/config/federation/bigquery/bq_ndt_server.sql#L19)):*
> ```SQLPL
>     AVG(IF(direction = "s2c", download, NULL)) AS value_download_avg_rate,
>     AVG(IF(direction = "c2s", upload, NULL)) AS value_upload_avg_rate,
>     APPROX_QUANTILES(IF(direction = "s2c", download, NULL), 2)[OFFSET(1)] as value_download_median_rate,
> ```

Hmm.  Just tried this with 99th and 99.9th.  The ratio of the two might be an interesting metric.  Closer to 1.0 probably means that we are seeing saturation of some kind.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/prometheus-support/131#-:-KzW04875pRQvKdpubCL:bnfp4nl)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
Thanks for the review!

---

Review status: 3 of 5 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[config/federation/bigquery/bq_ipv6_bias.sql, line 2 at r1](https://reviewable.io:443/reviews/m-lab/prometheus-support/131#-KzVu7qMOOHKSkwjC5FC:-KzZGpQ-8vibZiUIeGeB:b-896fix) ([raw file](https://github.com/m-lab/prometheus-support/blob/db0808c972271d905f392cb67008a04ce3647906/config/federation/bigquery/bq_ipv6_bias.sql#L2)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

typo: type *and* window scale
</blockquote></details>

Done.

---

*[config/federation/bigquery/bq_ipv6_bias.sql, line 13 at r1](https://reviewable.io:443/reviews/m-lab/prometheus-support/131#-KzVt_wEWRaf-pBkQ6CB:-KzZPUUk0Pwfb6NFksoZ:b-9g25gn) ([raw file](https://github.com/m-lab/prometheus-support/blob/db0808c972271d905f392cb67008a04ce3647906/config/federation/bigquery/bq_ipv6_bias.sql#L13)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Looks like 678 accounts for most of the counts.  Consider splitting into separate 6,7,8 counters?
</blockquote></details>

Yeah. Good idea. Split 6, 7, & 8.

---

*[config/federation/bigquery/bq_ipv6_bias.sql, line 25 at r1](https://reviewable.io:443/reviews/m-lab/prometheus-support/131#-KzVuLp8HE6E1kwANOVp:-KzZHE1C7X_jwgTpe-5o:b-4ht18z) ([raw file](https://github.com/m-lab/prometheus-support/blob/db0808c972271d905f392cb67008a04ce3647906/config/federation/bigquery/bq_ipv6_bias.sql#L25)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

The constraint seems a bit more subtle.  If you do this at 1:00 am, some of the data will be only 25 hours old.  Do you care?  I expect with the incremental scraper, it should be fine to run this at noon for the previous day, i.e., only 12 hour delay, right?
</blockquote></details>

Good question. I'm still learning tricks here.

Originally, I thought I was constrained by the refresh interval used by the bqe. But, I've added a calculation for the correct _PARTITIONTIME that is better than a static -2d. Basically, whenever it runs, _PARTITIONTIME will equal the most recent "completed" day. Where "completed" is 36h after start of day.

I've updated this for both queries.

---

*[config/federation/bigquery/bq_ipv6_bias.sql, line 31 at r1](https://reviewable.io:443/reviews/m-lab/prometheus-support/131#-KzVto3mVK05K4_XcfwQ:-KzZHsUtEu2QMuU-Bwpt:b-kufq3l) ([raw file](https://github.com/m-lab/prometheus-support/blob/db0808c972271d905f392cb67008a04ce3647906/config/federation/bigquery/bq_ipv6_bias.sql#L31)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

FYI - be aware that ORDER BY can cause memory problems with large number of items.  Definitely no problem here, but good to keep in mind.
</blockquote></details>

Good to know. Makes sense.

---

*[config/federation/bigquery/bq_ndt_server.sql, line 8 at r1](https://reviewable.io:443/reviews/m-lab/prometheus-support/131#-KzVuzeBfPPNwfAMzHng:-KzZZv7IyWAlGd8u6Y86:b-qc3xb5) ([raw file](https://github.com/m-lab/prometheus-support/blob/db0808c972271d905f392cb67008a04ce3647906/config/federation/bigquery/bq_ndt_server.sql#L8)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

I think that 99th percentile would probably be very useful as well.  Possibly 99.9th?

BTW, found this, which looks interesting:
https://cloud.google.com/blog/big-data/2016/10/a-google-sre-explores-github-reliability-with-bigquery

</blockquote></details>

Good question. I've added 99th & 99.9th percentile calculations and notes to evaluate their utility so we don't keep them if they're not helpful. the `AVG` calculation was also one of those in my head. I think we expect the median values to be stable over time. And, these queries are meant to be early warning signals for some major instrumentation regression from ndt or kernel upgrades.

---

*[config/federation/bigquery/bq_ndt_server.sql, line 12 at r1](https://reviewable.io:443/reviews/m-lab/prometheus-support/131#-KzVyirqrjI8fGUerTBg:-KzZIqJXcJuRRy8mXMyd:b-896fix) ([raw file](https://github.com/m-lab/prometheus-support/blob/db0808c972271d905f392cb67008a04ce3647906/config/federation/bigquery/bq_ndt_server.sql#L12)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

FYI, the #standardSQL only seems to be effective if it is the first line, so recommend putting it at the top.
</blockquote></details>

Done.

---

*[config/federation/bigquery/bq_ndt_server.sql, line 19 at r1](https://reviewable.io:443/reviews/m-lab/prometheus-support/131#-KzW-MW2uN7beHJTj_GC:-KzZ_TQtDViURISzLwjv:b-896fix) ([raw file](https://github.com/m-lab/prometheus-support/blob/db0808c972271d905f392cb67008a04ce3647906/config/federation/bigquery/bq_ndt_server.sql#L19)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Hmm.  Just tried this with 99th and 99.9th.  The ratio of the two might be an interesting metric.  Closer to 1.0 probably means that we are seeing saturation of some kind.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/prometheus-support/131#-:-Kz_VQw52TKkWca_txKA:b-9qdoz3)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
@mattmathis I need to push this out so we have some more data collected over the weekend. I'm not sure if you had a chance to look at the ipv6 bias query but I've changed the bins to better match the observed distribution of the current sites. We can easily change this in the future too.
